src/com/amazon/ionelement/api/IonElement.kt

```diff
@@ -110,145 +59,172 @@ interface IonElement {
-    val blobValueOrNull: IonByteArray? get() = expectNullOr(ElementType.BLOB).run { null }
+/**
+ * Represents an Ion symbol.
+ *
+ * Includes no additional functionality over [TextElement], but serves to provide additional type safety when
+ * working with elements that must be Ion symbols.
+ */
+interface SymbolElement : TextElement {
+    override fun copy(annotations: List<String>, metas: MetaContainer): SymbolElement
+}
 
-    /** See [IonElement]. */
-    val clobValue: IonByteArray get() = handleNull { clobValueOrNull }
+/** Represents an Ion clob or blob. */
+interface LobElement : IonElement {
+    val bytesValue:  IonByteArray
+    override fun copy(annotations: List<String>, metas: MetaContainer): LobElement
+}
 
-    /** See [IonElement]. */
-    val clobValueOrNull: IonByteArray? get() = expectNullOr(ElementType.CLOB).run { null }
+/**
+ * Represents an Ion blob.
+ *
+ * Includes no additional functionality over [LobElement], but serves to provide additional type safety when
+ * working with elements that must be Ion blobs.
+ */
+interface BlobElement : LobElement {
+    override fun copy(annotations: List<String>, metas: MetaContainer): BlobElement
+}
 
-    /** See [IonElement]. */
-    val containerValue: IonElementContainer get() = handleNull { containerValueOrNull }
+/**
+ * Represents an Ion clob.
+ *
+ * Includes no additional functionality over [LobElement], but serves to provide additional type safety when
+ * working with elements that must be Ion clobs.
+ */
+interface ClobElement : LobElement {
+    override fun copy(annotations: List<String>, metas: MetaContainer): ClobElement
+}
 
-    /** See [IonElement]. */
-    val containerValueOrNull: IonElementContainer? get() = expectNullOr(ElementType.LIST, ElementType.SEXP).run { null }
+/**
+ * Represents an Ion list, s-expression or struct.
+ *
+ * Items within [values] may or may not be in a defined order.  The order is defined for lists and s-expressions,
+ * but undefined for structs.
+ */
+interface ContainerElement : IonElement {
+    /** The number of values in this container. */
+    val size: Int
 
-    /** See [IonElement]. */
-    val listValue: IonElementContainer get() = handleNull { listValueOrNull }
+    val values: Iterable<AnyElement>
 
-    /** See [IonElement]. */
-    val listValueOrNull: IonElementContainer? get() = expectNullOr(ElementType.LIST).run { null }
+    override fun copy(annotations: List<String>, metas: MetaContainer): ContainerElement
+}
 
-    /** See [IonElement]. */
-    val sexpValue: IonElementContainer get() = handleNull { sexpValueOrNull }
+/**
+ * Represents an ordered collection element such as an Ion list or s-expression.
+ *
+ * Includes no additional functionality over [ContainerElement], but serves to provide additional type safety when
+ * working with ordered collection elements.
+ */
+interface SeqElement : ContainerElement {
+    override fun copy(annotations: List<String>, metas: MetaContainer): SeqElement
+}
```

Comment:
Shouldn't this have a `get` operator?

Reply:
It *could have one...

Reply:
The next question also is, if we support `seq[i]` and `seq.size`, why doesn't `SeqElement` also implement `List<IonElement>`?